### ISSUE TITLE
Guidelines for input/output handling

### ISSUE THREAD

**Issue by srmnitc** [pyiron/pyiron#1412]
This is closely related to #1403 

An important aspect in various projects is the incorporation of more and more software tools into pyiron. This calls for a streamlined method to handle input and output data. At the moment, the LAMMPS integration uses `GenericParameters`, while some of the newer integrations employ `DataContainer`. If `DataContainer` is the way forward, we should change the existing implementations.  

Would the ideal way to use `DataContainer` be:

```
self.input = DataContainer()
```

inside the job class? Sometimes validation/transformation of the input parameters are needed. If this is case, would the best choice be to make an input class derived from `DataContainer`?

```
class Input(DataContainer):
      ...
      @property
      def temperature(self):
          return self._temperature
      @temperature.setter
      def temperature(self, val):
          if isinstance(val, list):
              self._temperature = val

...
...

self.input = Input()
```

It might be a good idea to discuss what is the best/recommended way to do this. In addition, existing classes might need to be updated to be compatible with `DataContainer`. For example, the [LammpsPotential](https://github.com/pyiron/pyiron_atomistics/blob/master/pyiron_atomistics/lammps/potential.py#L27).

**Comment by pmrv** [comment]
At its heart `DataContainer` is a *generic* and *recursive* data structure, that also knows how to talk to HDF.  That means it's actually a bit overkill for most input classes for specific codes since they rarely need to be nested and if they do (e.g. in the `DamaskJob` from pyiron_continuum) the different levels end up meaning different things (exception: `Group` from sphinx). So input classes for specific codes are actually very rarely in the position that they also want to be recursive data structures and so should *not* derive from it directly since that would break the [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle).  Instead @liamhuber came up with `HasStorage`, which does not need to be a generic data structure, but still abstracts away the storage details of HDF for its sub classes.  Deriving directly would also have the side effect that users can by mistake store more attributes in input classes than the developer had wanted to allow (e.g. by simple typo).  For now this argument is only documented in the docstring of `DataContainer`, so we should move this also to a more formal spec like we tried to get started [here](https://github.com/pyiron/pyiron_base/tree/master/specs) (not very successfully yet).

So I'd rewrite your example closer to this
```python
class Input(HasStorage):
      ...
      @property
      def temperature(self):
          return self.storage.temperature
      @temperature.setter
      def temperature(self, val):
          if isinstance(val, list):
              self.storage.temperature = val

...
...

self.input = Input()
```

I can see that this can become a bit tiresome to write after a while though.  In my own code I'm currently prototyping something along the lines of `namedtuple`, i.e. we would provide a function that creates the `HasStorage` sub class and properties for you.

As a quick sketch

```python
def make_property(name, type=object, default=None, encode=lambda x: x, decode=lambda x: x, doc=None):
    def fget(self):
        if name not in self.storage:
            return default
        else:
            return decode(self.storage[name])
    def fset(self, value):
        if not isinstance(value, type):
            raise TypeError(f'{name} must be of type {type}!')
        self.storage[name] = encode(value)
    return property(fget=fget, fset=fset, doc=doc)
```

I've added `encode`/`decode` because it might be useful to only store smaller information in HDF that can be used to recreate the full object later.  I use this when saving reference jobs for example.  The input really only needs to store the job id and can transparently load the full job when you access the input.

You could use this directly

```python
class MyInput(HasStorage):
    temperature = make_property('temperature', type=list, default=[], doc='My Fancy Temperature List')
    structure = make_property('structure', type=Atoms, default=None, doc='My Structure')
```

or we go one step further

```python
def make_input(name, **attrs):
    attrs = {k: make_property(name=k, type=v) for k, v in attrs.items()}
    return type(name, (HasStorage,), attrs)
```

but that could be a bit unwieldy if we want to support a lot of options to `make_property`.

Anyways, I think it's a very important topic to discuss and we should flesh out a spec for this in a pyiron meeting.

**Comment by liamhuber** [comment]
I like Marvin's sketches for how to avoid the overkill that comes with simply using DataContainer, but I have to admit I have frequently just over killed in the past 😝

One aspect I like about this entire direction is it allows us to begin constructing constructing our input in an OOP manner -- which has a great side effect of having consistent naming conventions.

A closely related issue is how we break apart representations and simulations. ie lammps needs a potential while vasp needs kpoints, but md needs temperature while minimization needs convergence limit. I think there are great opportunities to streamline the code by doing loads of construction by composition, but it will require being a bit intentional about the process.

**Comment by liamhuber** [comment]
Ps sorry no code styling, I haven't figured out how to get that working from my mobile app yet. 'Code' hasn't done it in the past

**Comment by niklassiemer** [comment]
I like and support the idea as well :) ps for `code` style you need the backwards `

**Comment by pmrv** [comment]
> One aspect I like about this entire direction is it allows us to begin constructing constructing our input in an OOP manner -- which has a great side effect of having consistent naming conventions.

Yes, in my experiments that works pretty well.

**Comment by liamhuber** [comment]
> I like and support the idea as well :) ps for `code` style you need the backwards ` 

Yeah, my phone keyboard has forward single quote, but not single backwards quote 🙁

**Comment by pmrv** [comment]
After our quick session in the pyiron meeting today, I got it down to this MWE.

```python
from pyiron_base.generic.object import HasStorage

class InputField:

    def __init__(
            self,
            fget, fset,
            type, doc,
    ):
        self.fget = fget
        self.fset = fset
        self.__doc__ = doc
        self._type = type

    def __get__(self, instance, owner=None):
        value = instance.storage.get(self._name, default=None)
        if value is None:
            value = self._default
        if value is None:
            raise AttributeError(f'"{self._name}" not defined (yet)!')
        if self._getter is not None:
            value = self.fget(instance, value)
        return value

    def __set__(self, instance, value):
        if self._setter is not None:
            value = self.fset(instance, value)
        instance.storage[self._name] = value


def make_scalar_input(
        name, type, doc,
        default = None,
        encode = lambda x: x,
        decode = lambda x: x
):
    def fget(self) -> type:
        if name not in self.storage:
            if default is not None:
                return default
            else:
                raise AttributeError(f"'{name}' not defined yet!")
        else:
            return decode(self.storage[name])
    def fset(self, value: type):
        if not isinstance(value, type):
            raise TypeError(f'{name} must be of type {type}!')
        self.storage[name] = encode(value)
    prop = InputField(fget=fget, fset=fset, type=type, doc=doc)
    return prop

class Input(HasStorage):

    def __init__(self):
        super().__init__()
        self.__doc__ = self._make_doc()

    def _make_doc(self):
        cn = self.__class__.__name__
        doc = cn + "\n" + "-"*len(cn) + "\n" + self.__doc__ + "\n\n" \
                + "Attributes:\n" \
                + "-----------"
        for k, v in self.__class__.__dict__.items():
            if isinstance(v, InputField):
                doc += f"\n\t{k} ({v._type}): {v.__doc__}"
        return doc

class MDInput(Input):
    """Input parameters for MD calculations."""

    temperature = make_scalar_input(
            "temperature", float, doc="Run at this temperature"
    )
    steps = make_scalar_input(
            "steps", int, doc="How many steps to integrate"
    )
    timestep = make_scalar_input(
            "timestep", float, default=1e-15, doc="Time step for the integration in fs"
    )
```
then doing
```python
md = MDInput()
md?
```
in an IPython shell gives you
```
Type:            MDInput
String form:     <pyiron_base.generic.input.MDInput object at 0x7fb1808ffca0>
File:            ~/science/phd/dev/pyiron_base/pyiron_base/generic/input.py
Docstring:
MDInput
-------
Input parameters for MD calculations.

Attributes:
-----------
        temperature (<class 'float'>): Run at this temperature
        steps (<class 'int'>): How many steps to integrate
        timestep (<class 'float'>): Time step for the integration in fs
Class docstring: Input parameters for MD calculations.
```

**Comment by samwaseda** [comment]
> in an IPython shell gives you
> 
> ```
> Type:            MDInput
> String form:     <pyiron_base.generic.input.MDInput object at 0x7fb1808ffca0>
> File:            ~/science/phd/dev/pyiron_base/pyiron_base/generic/input.py
> Docstring:
> MDInput
> -------
> Input parameters for MD calculations.
> 
> Attributes:
> -----------
>         temperature (<class 'float'>): Run at this temperature
>         steps (<class 'int'>): How many steps to integrate
>         timestep (<class 'float'>): Time step for the integration in fs
> Class docstring: Input parameters for MD calculations.
> ```

This looks amazing! In addition, I guess it makes sense to include the default values in the docstring as well, so the last one would be: `timestep (<class 'float'>): Time step for the integration in fs (default: 1e-15)`.

**Comment by samwaseda** [comment]
> ```python
> class MDInput(Input):
>     """Input parameters for MD calculations."""
> 
>     temperature = InputField(
>             name="temperature", type=float, doc="Run at this temperature"
>     )
>     steps = InputField(
>             name="steps", type=int, doc="How many steps to integrate"
>     )
>     timestep = InputField(
>             name="timestep", type=float, default=1e-15, doc="Time step for the integration in fs"
>     )
> ```

While working on this, I realised that the property itself becomes `InputField`, i.e.:

```python
md.temperature
```

returns `InputField` instead of its actual value, correct? If I remember @liamhuber's comment well, it's actually difficult to make it return the value, but do you see a way around it?

**Comment by samwaseda** [comment]
Sorry my mistake, I just realised that that's what's done inside `__get__`

**Comment by samwaseda** [comment]
Now I implemented it in [this PR](https://github.com/pyiron/pyiron_base/pull/753), because I didn't see it so far implemented, with a few modifications.
